packages/compass-preferences-model/src/compass-web-preferences-access.ts

```diff
@@ -12,6 +12,16 @@ export class CompassWebPreferencesAccess implements PreferencesAccess {
       logger: createNoopLogger(),
       preferencesStorage: new InMemoryStorage(preferencesOverrides),
     });
+
+    if (process.env.E2E_TEST_CLOUD_WEB_ENABLE_PREFERENCE_SAVING) {
+      // Useful for e2e test to override preferences.
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (globalThis as any).__compassWebE2ETestSavePreferences = async (
+        attributes: Partial<AllPreferences>
+      ) => {
+        await this._preferences.savePreferences(attributes);
+      };
+    }
```

Comment:
We don't expect this class to be a singleton, but this code make is behave like one. This also means that explicit e2e code, variables that can be set when it's running in mms, gets included as part of the constructor logic being used in prod. Let's try to find a way to provide this interface without hardcoding e2e var checks inside a class constructor.

One suggestion would be to do it similarly to a sandbox provider in the connection storage. We can use it to provide some sort of way to register a preferences update listener that sandbox would set up without the need for this to be done directly in the preferences class. Roughly a hook like this that would handle preferences instance creation:

```tsx
function useCompassWebPreferences() {
  const preferencesAccess = // ... current logic to set up preferences class as a ref
  const onPreferencesUpdateTriggered = useSandboxPreferencesUpdateTrigger();
  useEffect(() => {
    // Provided by the sandbox, registers this listener through global callback that you implemented, returns some sort of cleanup if needed
    return onPreferencesUpdateTriggered?.(preferences => {
      preferencesAccess.current.savePreferences()
    })
  }, [])
  return preferencesAccess
}
```

That way this code would never be exposed as an interface on the compass-web component or otherwise would allow to be activated in production, and would allow us to set up something in the sandbox for testing purposes.

Reply:
Great suggestion! Thanks.